### PR TITLE
chapter14_following_users

### PR DESCRIPTION
**Pull request by tenyiting**
chapter14_following_users

### CODE REVIEW

**Comment by tenyiting** [comment]
ready

**Comment by tenyiting** [comment]
ready

File: app/models/user.rb [hunk 1]
```diff
@@ -2,15 +2,35 @@ class User < ApplicationRecord
...
+    foreign_key: :followed_id, dependent: :destroy
+  has_many :following, through: :active_relationships, source: :followed
+  has_many :followers, through: :passive_relationships, source: :follower
 
-  validates :name, presence: true
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
-  validates :email, presence: true, length: {maximum: 255}, format: {with: VALID_EMAIL_REGEX}
-  validates :email, uniqueness: true
+
+  before_save {self.email = email.downcase}
```
Comment (huongnt-2545) [inline app/models/user.rb]: remove this line, it's same with line 3

File: app/models/user.rb [hunk 2]
```diff
@@ -2,15 +2,35 @@ class User < ApplicationRecord
   attr_accessor :remember_token, :activation_token, :reset_token
   before_save :downcase_email
   before_create :create_activation_digest
+  has_secure_password
```
Comment (huongnt-2545) [inline app/models/user.rb]: move this line place to after validate area
ex:
```
# something
validates :password, presence: true, length: {minimum: 6}, allow_nil: true

has_secure_password

# something
```

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -6,7 +6,7 @@
     <%= csrf_meta_tags %>
     <%= csp_meta_tag %>
     <%= stylesheet_link_tag "application", media: 'all', "data-turbolinks-track": "reload" %>
-    <%= javascript_tag "application", "data-turbolinks-track": "reload" %>
+    <%= javascript_path "application", "data-turbolinks-track": "reload" %>
```
Comment (huongnt-2545) [inline app/views/layouts/application.html.erb]: why so you change this?

File: config/routes.rb [hunk 4]
```diff
@@ -1,23 +1,27 @@
 Rails.application.routes.draw do
   scope "(:locale)", locale: /#{I18n.available_locales.join("|")}/ do
+    get 'password_resets/new'
```
Comment (huongnt-2545) [inline config/routes.rb]: remove line 3, 4

File: test/fixtures/relationships.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (huongnt-2545) [inline test/fixtures/relationships.yml]: delete this file

File: test/models/relationship_test.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (huongnt-2545) [inline test/models/relationship_test.rb]: delete this file

File: app/views/layouts/application.html.erb (hunk 3 above)
Comment (tenyiting) [reply app/views/layouts/application.html.erb]: test something & forgot to change back

**Comment by tenyiting** [comment]
ready

File: app/views/users/_unfollow.html.erb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<%= form_for(current_user.active_relationships
+.find_by(followed_id: @user.id),
+html: { method: :delete }, remote: true) do |f| %>
+<%= f.submit "Unfollow", class: "btn" %>
```
Comment (ghost) [inline app/views/users/_unfollow.html.erb]: indent please

File: app/views/layouts/_header.html.erb [hunk 8]
```diff
@@ -24,8 +24,8 @@
         <li><%= link_to t(:log_in), login_path %></li>
       <% end %>
 
-      # <li><%= link_to t(:c_en), locale: :en %></li>
-      # <li><%= link_to t(:c_ja), locale: :ja %></li>
+      <li><%= link_to t(:c_en), locale: :en %></li>
+      <li><%= link_to t(:c_ja), locale: :ja %></li>
```
Comment (ghost) [inline app/views/layouts/_header.html.erb]: indent please

File: app/views/layouts/_header.html.erb [hunk 9]
```diff
@@ -3,8 +3,8 @@
     <%= link_to t(:h_title) , "#", id: "logo" %>
     <nav>
       <ul class="nav navbar-nav navbar-right">
-      <li><%= link_to t(:Home), home_path %></li>
-      <li><%= link_to t(:Help), help_path %></li>
+      <li><%= link_to t(:home), home_path %></li>
+      <li><%= link_to t(:help), help_path %></li>
```
Comment (ghost) [inline app/views/layouts/_header.html.erb]: indent pleaes

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -31,22 +31,26 @@ def remember user
...
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+
   def forget(user)
-    user.forget
-    cookies.delete(:user_id)
-    cookies.delete(:remember_token)
+   user.forget
+   cookies.delete(:user_id)
+   cookies.delete(:remember_token)
```
Comment (ghost) [inline app/helpers/sessions_helper.rb]: Wrong indent, need 2 space please

**Review by ghost — COMMENTED** [review]
cmt

**Comment by tenyiting** [comment]
ready

**Comment by tenyiting** [comment]
ready

**Comment by tenyiting** [comment]
ready

File: app/views/shared/_stats.html.erb [hunk 11]
```diff
@@ -0,0 +1,15 @@
+<% @user ||= current_user %>
+<div class="stats">
+  <a href="<%= following_user_path(@user) %>">
+    <strong id="following" class="stat">
+      <%= @user.following.count %>
+    </strong>
+    following
```
Comment (ghost) [inline app/views/shared/_stats.html.erb]: i18n

File: app/views/shared/_stats.html.erb [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+  </a>
+  <a href="<%= followers_user_path(@user) %>">
+    <strong id="followers" class="stat">
+      <%= @user.followers.count %>
+    </strong>
+    followers
```
Comment (ghost) [inline app/views/shared/_stats.html.erb]: i18n

File: app/models/user.rb [hunk 13]
```diff
@@ -23,7 +43,7 @@ def digest string
       else
         BCrypt::Engine.cost
       end
-        BCrypt::Password.create string, cost: cost
+       BCrypt::Password.create string, cost: cost
```
Comment (ghost) [inline app/models/user.rb]: align BCrypt with end at line 45

**Review by ghost — COMMENTED** [review]
cmt

**Comment by tenyiting** [comment]
ready

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -57,6 +58,20 @@ def destroy
     redirect_to users_url
   end
 
+  def following
+    @title = "Following"
```
Comment (ghost) [inline app/controllers/users_controller.rb]: use i18n

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -57,6 +58,20 @@ def destroy
...
+    @user  = User.find_by(id: params[:id])
+    @users = @user.following.paginate(page: params[:page])
+    render "show_follow"
+  end
+
+  def followers
+    @title = "Followers"
```
Comment (ghost) [inline app/controllers/users_controller.rb]: use i18n

**Review by ghost — COMMENTED** [review]
small comment

**Comment by tenyiting** [comment]
ready

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -57,6 +58,20 @@ def destroy
     redirect_to users_url
   end
 
+  def following
+    @title: t(:following)
```
Comment (ghost) [inline app/controllers/users_controller.rb]:
```suggestion
    @title = t(:following)
```

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -57,6 +58,20 @@ def destroy
...
+    @user  = User.find_by(id: params[:id])
+    @users = @user.following.paginate(page: params[:page])
+    render "show_follow"
+  end
+
+  def followers
+    @title: t(:followers)
```
Comment (ghost) [inline app/controllers/users_controller.rb]:
```suggestion
    @title = t(:followers)
```

**Review by ghost — COMMENTED** [review]
cmt

**Comment by tenyiting** [comment]
ready

**Review by ghost — COMMENTED** [review]
merrged

**Pull request merged**